plugins/parquet-store/parquet_store.cpp

```diff
@@ -0,0 +1,722 @@
+    },
+    [&](const arrow::Scalar&) -> void {},
+  };
+  caf::visit(f, *max_value);
+  return slice;
+}
+
+// Handler for `vast::query` that is shared between active and passive stores.
+// Returns a the number of events that match the query.
+template <typename Actor>
+caf::expected<uint64_t>
+handle_lookup(Actor& self, const vast::query& query,
+              const std::vector<table_slice>& table_slices) {
+  if (table_slices.empty())
+    return 0;
+  // table slices from parquet can't utilize query hints because we don't retain
+  // the global ids.
+  const auto& ids = vast::ids{};
+  auto expr = expression{};
+  if (auto e = tailor(query.expr, table_slices[0].layout()); e) {
+    expr = *e;
+  }
+  uint64_t num_hits = 0ull;
+  auto handle_query = detail::overload{
+    [&](const query::count& count) -> caf::expected<uint64_t> {
+      VAST_ASSERT(count.mode != query::count::estimate);
+      std::shared_ptr<arrow::RecordBatch> batch{};
+      for (const auto& slice : table_slices) {
+        auto result = count_matching(slice, expr, ids);
+        num_hits += result;
+        self->send(count.sink, result);
+      }
+      return num_hits;
+    },
+    [&](const query::extract& extract) -> caf::expected<uint64_t> {
+      for (const auto& slice : table_slices) {
+        auto final_slice = filter(slice, expr, ids);
+        if (final_slice) {
+          num_hits += final_slice->rows();
+          self->send(extract.sink, *final_slice);
+        }
+      }
+      return num_hits;
+    },
+  };
+  return caf::visit(handle_query, query.cmd);
+}
+
+std::shared_ptr<arrow::Schema> parse_arrow_schema_from_metadata(
+  const std::shared_ptr<parquet::FileMetaData>& parquet_metadata) {
+  if (!parquet_metadata)
+    return {};
+  auto arrow_metadata
+    = parquet_metadata->key_value_metadata()->Get("ARROW:schema");
+  if (!arrow_metadata.ok())
+    return {};
+  auto decoded = detail::base64::decode(*arrow_metadata);
+  auto schema_buf = arrow::Buffer(decoded);
+  arrow::ipc::DictionaryMemo dict_memo;
+  arrow::io::BufferReader input(schema_buf);
+  auto arrow_schema = arrow::ipc::ReadSchema(&input, &dict_memo);
+  if (!arrow_schema.ok())
+    return {};
+  return *arrow_schema;
+}
+
+caf::expected<std::shared_ptr<arrow::Table>>
+read_parquet_buffer(const chunk_ptr& chunk) {
+  VAST_ASSERT(chunk);
+  auto bufr = std::make_shared<arrow::io::BufferReader>(as_arrow_buffer(chunk));
+  auto parquet_reader = parquet::ParquetFileReader::Open(bufr);
+  auto arrow_schema
+    = parse_arrow_schema_from_metadata(parquet_reader->metadata());
+  std::unique_ptr<parquet::arrow::FileReader> file_reader{};
+  if (!parquet::arrow::OpenFile(bufr, arrow::default_memory_pool(),
+                                &file_reader)
+         .ok())
+    return caf::exit_reason::unhandled_exception;
+  std::shared_ptr<arrow::Table> table{};
+  if (!file_reader->ReadTable(&table).ok())
+    return caf::exit_reason::unhandled_exception;
```

Comment:
This looks like the wrong error code to me in both cases. Can we use something fitting from `vast::ec`, and add the stringified status to the error?

Reply:
The arrow documentation is rather sparse on the possible error conditions. I chose `vast::ec::parse_error`, which hopefully catches the typical stuf happening when opening and reading parquet files (the bytes are already clear as the're coming from the file system actor).